### PR TITLE
Loess Regression

### PR DESCRIPTION
**Pull request by dvanamstel**
Still WIP, but we think in an enough mature stage to have a review and actually being a useful tool.

Prio 1 is to add tests, Prio 2 is to clean the _get_knn_symmetric_indices windowing function.

PS. My git notes the file docs/_static/js/modernizr.min.js has changed (not intentional), is this one automatically updated because there are new Docstrings in the package? Left it out of the commits, let me know if we should add.

### CODE REVIEW

File: sklego/loess.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+
+import matplotlib.pyplot as plt
+import numpy as np
+from sklearn.base import BaseEstimator
+from sklearn.base import TransformerMixin
+from sklearn.linear_model import LinearRegression
+from sklearn.neighbors import NearestNeighbors
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import PolynomialFeatures
+
```
Comment (koaning) [inline sklego/loess.py]: `Loess` can be part of `sklego.linear`.

File: sklego/loess.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+                 model=None,
+                 n_degree=2,
+                 transformer_type=True,
+                 window_method='fixed',
+                 window_size=2,
+                 step_size=.2,
+                 fraction=.1):
+        """
+
+        Loess Regression. This class implements multiple methods of Loess regression for the single
```
Comment (koaning) [inline sklego/loess.py]: image

not that i want to make this a *super* important thing, but could you make the width a bit smaller of the docstring here? makes reviewing in smaller screens easier.

File: sklego/loess.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+
+        In transformer mode, the instance will take each data_point and return it's y value changed
+        according to the Loess regression. Hence, no interpolation is done in between missing
+        datapoints.
+
+        In estimator mode, the instance will create a x_focal_base to evalue the y values for. The
+        created x_focal_base is an equally spaced distance vector between x_min and x_man with
+        distance step_size between elements. Hence, interpolation will happen because of the fixed
+        step sizes.
+
```
Comment (koaning) [inline sklego/loess.py]: might be good to add a reference to the paper in the docstring as well.

File: sklego/loess.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            x_window = x[self.indices[index]].reshape(-1, 1)
+            y_window = y[self.indices[index]].reshape(-1, 1)
+
+            # TODO: implement iterative weighted fit
+            model = self.model.fit(x_window, y_window)
+            self.y_focal_base = np.append(self.y_focal_base, model.predict(x_focal))
+            self.model_per_window = np.append(self.model_per_window, model)
+
+
+def plot_windows(x, y, x_focal_base, y_focal_base, indices):
```
Comment (koaning) [inline sklego/loess.py]: if you want to make plots and have this be part of the library -> lets discuss that but at least place it in a submodule for it `sklego.plotting` for example.

File: sklego/loess.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        ax = plt.axes(xlim=(np.min(x), np.max(x)), ylim=(np.min(y), np.max(y)))
+        ax.scatter(x[indices[index]], y[indices[index]])
+        ax.scatter(x_focal_base[index], y_focal_base[index], c='r')
+        ax.set_title(f'Window for {x_focal_base[index]}')
+        plt.pause(0.01)
+        plt.show()
+
+
+# HELPER FUNCTIONS TO GENERATE RANDOM DATA
+def random_x(minimum_val, maximum_val, size):
```
Comment (koaning) [inline sklego/loess.py]: this should be part of `sklego.datasets` if we think it is important enough to have.

File: sklego/loess.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+    Generate n random data-points of size between minimum_val and right bounds
+    :param minimum_val: float, minimal value of the generated data
+    :param maximum_val: float, maximum value of the generated data
+    :param size: tuple or list, shape of the desired output
+    :return: x, y
+    """
+    return (maximum_val - minimum_val) * np.random.random(size=size) + minimum_val
+
+
+def generate_noisy_sine_data(noise_std):
```
Comment (koaning) [inline sklego/loess.py]: this should be part of `sklego.datasets` if we think it is important enough to have.

File: sklego/loess.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+            # TODO: implement iterative weighted fit
+            model = self.model.fit(x_window, y_window)
+            self.y_focal_base = np.append(self.y_focal_base, model.predict(x_focal))
+            self.model_per_window = np.append(self.model_per_window, model)
+
+
+def plot_windows(x, y, x_focal_base, y_focal_base, indices):
+    """
+    TODO: Use matplotlib.animation to create video
```
Comment (koaning) [inline sklego/loess.py]: if we want to make this animation availble it might be better to place this somewhere in the documentation. perhaps add a script to a new `/examples` folder in root?

**Comment by koaning** [comment]
hi django, 

feel free to notify me again when you believe this work is ready for a review. if you still need to add lots of things it might be better to delay the review. i prefer to see a few tests before i start properly reviewing.

in the meantime to answer your questions: 

1. yes testing should happen first
2. cleaning code is always good
3. feel free to ignore anything related to the docs, that's been patched. you may want to update your branch such that it has all the changes that master currently has (your branch is running a bit behind now i think).

**Comment by koaning** [comment]
to make the reviewing a bit easier, feel free to add some screenshots of what your method does. check out [this pr](https://github.com/koaning/scikit-lego/pull/91) to see an example of what i mean.

**Comment by koaning** [comment]
feel free to check if flake8 passes on your local machines.

**Comment by koaning** [comment]
**discussion** 

so ... i've also made a hacky implementation of a loess prediction today. it looks like this: 

```
def make_prediction(x_point):
    np_arr = np.array([x_point])
    wts = np.exp(-np.array([euclidean(x, np_arr) for x in X])/10)
    return LinearRegression().fit(X, y, sample_weight=wts).predict(np_arr.reshape(-1, 1)).reshape(-1)

make_prediction([1, 2])
``` 

it is super hacky. but substantially it is a short hack and it has a few benefits:

- it is easy to maintain
- if a span is given we can first make a subselect on `X` to accomodate a span
- the `wts` param can be used to allow for any kernel 

im havinig a bit of trouble fully understand the code that was submitted. is there a particular reason why the approach listed here won't work?

File: sklego/loess.py (hunk 1 above)
Comment (dvanamstel) [reply sklego/loess.py]: Considering:
- Loess can use non-linear models for the local fits
- The local model fitting is non-linear in itself

Are you sure about this?

File: sklego/loess.py (hunk 5 above)
Comment (dvanamstel) [reply sklego/loess.py]: I guess the functions itself are not that important, will move them to the development notebook.

**Comment by dvanamstel** [comment]
> im havinig a bit of trouble fully understand the code that was submitted. is there a particular reason why the approach listed here won't work?

TL;DR: The current implementation has a lot of code for creating the animation, this can be moved outside of the actual LoessSmoother. Let's discuss on implementation and then indeed Tomi and I can revise the whole thing before the next review.


---
I don't see any reason why your proposal wouldn't work. Actually, if you look at the last _fit_model_per_window function, it is similar to your code and the core of the whole thing.

All other code around it mostly to create and store the information to create the GIF you posted [here](https://github.com/koaning/scikit-lego/issues/3#issuecomment-468754430).

I do agree it is better to separate the actual loess estimator and creating fancy animations. I also think it is cleaner to use the wts param for windowing instead of the explicit dataset filtering.

If you want to integrate the Smoother in a sklearn pipeline I think two things can be considered:
1. When calling .fit, no actually fitting is done in the LoessSmoother. When .predict is called, under the hood it first fits the model.
2. When calling .fit, it actually creates and fits a model for all the windows in the dataset (the number of windows are defined by the window_size, step_size, etc. parameters). Then when calling predict, it uses the closest model for prediction.

Now that I think about it, I silently assumed and went for the second option. I currently can't think of a reason why the first wouldn't work.

I think it's best to have a discussion next to a whiteboard on what is required/desired, after which we can update stuff accordingly!

File: sklego/loess.py (hunk 1 above)
Comment (koaning) [reply sklego/loess.py]: actually ... we need to follow the sklearn api so it should be `sklego.linear_model` 

the local model *is* linear, it might just have polynomial features going in. the *features* are non-linear such that the model can be.

**Comment by koaning** [comment]
it seems like you've made a very small change to `modernizer.js`. if you could revert the change it might make merge-ing a whole lot easier. 

image
